net: lib: lwm2m_client_utils: Add support for connection pre-evaluation

juhaylinen opened the pull request:
Connection pre-evaluation estimates the LTE radio signal state and energy efficiency in a cell that the modem would use for data transmission. Application can use the information to decide when to start the data transmission.
If estimated energy usage is too high the application pauses the lwm2m client and waits for better signal state.
User can set evaluation period in milliseconds and maximum time in seconds to pause the transmission.

juhaylinen commented:
This PR needs to wait for Zephyr upmerge to pass CI tests.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-10520
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

rlubos commented on subsys/net/lib/lwm2m_client_utils/Kconfig:
```diff
@@ -248,6 +248,15 @@ config LWM2M_CLIENT_UTILS_RAI
 	  expected in the near future and the connection may be released.
 
+config LWM2M_CLIENT_UTILS_CONNEVAL
+	bool "Enable connection pre-evaluation"
```
Please add `[EXPERIMENTAL]` suffix to the option description.

rlubos commented on subsys/net/lib/lwm2m_client_utils/Kconfig:
```diff
@@ -248,6 +248,15 @@ config LWM2M_CLIENT_UTILS_RAI
 	  When AS RAI is configured, device may indicate that no further data is
 	  expected in the near future and the connection may be released.
 
+config LWM2M_CLIENT_UTILS_CONNEVAL
```
I wonder if this option/API should include `LTE` keyword, like `LTE_CONNEVAL`. With other networking technologies on the table now, current name might be too generic.

juhaylinen commented:
Refactored the code as discussed offline with @SeppoTakalo 

- Removed own work queue implementation and use delayed work instead.
- Use asynchronous AT command for connection pre-evaluation when evaluation is done multiple times. LTE link controller API may block too long for system work queue.

VeijoPesonen commented on subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_conneval.c:
```diff
@@ -0,0 +1,176 @@
+static struct lte_lc_conn_eval_params params;
+static enum lte_lc_energy_estimate energy_estimate;
+static uint64_t maximum_delay;
+static uint64_t poll_period;
```
```suggestion
static uint64_t poll_period; /* [ms] */
```

VeijoPesonen reviewed (changes requested):
Nothing major, generally looks good to me.

VeijoPesonen reviewed (approved):
Looks good to me

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-10520/8/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-10520/8/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-ci-nrfconnect-boot-fw-update | X | | |
| test-fw-nrfconnect-nrf-iot_zephyr_lwm2m | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-10520/8/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*
